src/main/java/org/apache/maven/plugins/checkstyle/CheckstyleViolationCheckMojo.java

```diff
@@ -605,70 +610,118 @@ private void checkDeprecatedParameterUsage( Object parameter, String name, Strin
         }
     }
 
-    private int countViolations( XmlPullParser xpp )
+    private List<Violation> getViolations( XmlPullParser xpp )
         throws XmlPullParserException, IOException
     {
-        int count = 0;
-        int ignoreCount = 0;
-        List<RuleUtil.Matcher> ignores = violationIgnore == null ? Collections.<RuleUtil.Matcher>emptyList()
-                        : RuleUtil.parseMatchers( violationIgnore.split( "," ) );
+        List<Violation> violations = new ArrayList<>();
 
         String basedir = project.getBasedir().getAbsolutePath();
         String file = "";
+
         for ( int eventType = xpp.getEventType(); eventType != XmlPullParser.END_DOCUMENT; eventType = xpp.next() )
         {
             if ( eventType != XmlPullParser.START_TAG )
             {
                 continue;
             }
-            else if ( "file".equals( xpp.getName() ) )
+
+            if ( "file".equals( xpp.getName() ) )
             {
                 file = PathTool.getRelativeFilePath( basedir, xpp.getAttributeValue( "", "name" ) );
-                //file = file.substring( file.lastIndexOf( File.separatorChar ) + 1 );
+                continue;
             }
-            else if ( "error".equals( xpp.getName() ) )
+
+            if ( !"error".equals( xpp.getName() ) )
             {
-                String severity = xpp.getAttributeValue( "", "severity" );
+                continue;
+            }
 
-                if ( !isViolation( severity ) )
-                {
-                    continue;
-                }
+            String severity = xpp.getAttributeValue( "", "severity" );
+            String source = xpp.getAttributeValue( "", "source" );
+            String line = xpp.getAttributeValue( "", "line" );
+            @Nullable String column = xpp.getAttributeValue( "", "column" );
+            String message = xpp.getAttributeValue( "", "message" );
+            String rule = RuleUtil.getName( source );
+            String category = RuleUtil.getCategory( source );
+
+            Violation violation = new Violation(
+                source,
+                file,
+                Integer.parseInt( line, 10 ),
+                severity,
+                message,
+                rule,
+                category
+            );
+            if ( column != null )
+            {
+                violation.setColumn( Integer.parseInt( column, 10 ) );
+            }
+
+            violations.add( violation );
+        }
+
+        return Collections.unmodifiableList( violations );
```

Comment:
worth javadocing that this method returns and the next method takes an unmodifiable list, even for private methods. This has burned me in the past since there's no info in the type signature.

Reply:
It is a good habit to never modify lists you did not create yourself. I never ran into this issue, because `.stream().filter().collect(toList())` will just create a new one. I think it should be worth mentioning if a list is mutable (that's the more suprising case). See the JAX-RS docs for example. If you implement a `ContainerRequestFilter`, the docs about the `ContainerRequestContext.getHeaders()` method will explicitly mention it will return a mutable list.

I can add a test if you like.

Reply:
The List interface is defined as a mutable data type. That might have been the wrong definition way back in Java 1.1, but we're not going to change it here. Developers do expect they can invoke the various methods on this class including set and friends without getting an UnsupportedOperationException, so if we're going to violate that expectation a warning is useful.

Reply:
Yes, Java 1.1. By now everyone knows that modifying a list unless explicitly mentioned is a bad design.

> Developers do expect they can invoke the various methods on this class including set and friends without getting an UnsupportedOperationException.

I disagree. I would never expect that. On the contrary, if I used them and did not get such an exception, I would expect various side effects. A copy of a list is almost always cheap. If it is the only copy and the orignal list is not used again, it is cheap to garbage collect.

Reply:
Trust me, everyone does not know that. I read lots of code, and nope, that principle is not nearly as commonly understood as it should be.

Reply:
Sounds you agree to my POV, but have fear that you as a reviewer might oversee that at some later time.
Feel free to change it yourself, because it is not a functional change. Other reason why I do not want to change it myself:
If some other method pops up, modifying that lust, the prints may be missing violations. I do not want to be the reason for that.

Reply:
> Trust me, everyone does not know that. 

That's not a good argument. In fact it's not an argument at all.

> I read lots of code.

That's not an argument either. Also, I also read lots of code.

Trust me, you'll see UnsupportedOpperationException in a unit test right away, while concurrent modification will hide like a Heisenberg bug.

Feel free to change it. Embrace side effect programming! 👍🏻

Reply:
Commited, now is a modifiable collection b/c of existing coding conventions.